src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -467,58 +489,113 @@ private static Expression AddResponseWritingToMethodCall(Expression methodCall,
                 return Expression.Lambda<Func<object?, HttpContext, Task>>(
                     responseWritingMethodCall, TargetExpr, HttpContextExpr).Compile();
             }
 
-            // We need to generate the code for reading from the body before calling into the delegate
+            // We need to generate the code for reading from the body or form before calling into the delegate
             var invoker = Expression.Lambda<Func<object?, HttpContext, object?, Task>>(
                 responseWritingMethodCall, TargetExpr, HttpContextExpr, BodyValueExpr).Compile();
 
             var bodyType = factoryContext.JsonRequestBodyType;
-            object? defaultBodyValue = null;
+            object? defaultContentValue = null;
 
-            if (factoryContext.AllowEmptyRequestBody && bodyType.IsValueType)
+            if (!factoryContext.ReadForm && factoryContext.AllowEmptyRequestBody && bodyType is { IsValueType: true })
             {
-                defaultBodyValue = Activator.CreateInstance(bodyType);
+                defaultContentValue = Activator.CreateInstance(bodyType);
             }
 
             return async (target, httpContext) =>
             {
-                object? bodyValue = defaultBodyValue;
+                object? contentValue = defaultContentValue;
+
+                var feature = httpContext.Features.Get<IHttpRequestBodyDetectionFeature>();
+                if (feature?.CanHaveBody == true)
+                {
+                    if (bodyType is not null)
+                    {
+                        try
+                        {
+                            contentValue = await httpContext.Request.ReadFromJsonAsync(bodyType);
+                        }
+                        catch (IOException ex)
+                        {
+                            Log.RequestBodyIOException(httpContext, ex);
+                            return;
+                        }
+                        catch (InvalidDataException ex)
+                        {
+                            Log.RequestBodyInvalidDataException(httpContext, ex);
+                            httpContext.Response.StatusCode = 400;
+                            return;
+                        }
+                    }
+                    else
+                    {
+                        try
+                        {
+                            contentValue = await httpContext.Request.ReadFormAsync();
+                        }
+                        catch (IOException ex)
+                        {
+                            Log.RequestFormIOException(httpContext, ex);
+                            return;
+                        }
+                        catch (InvalidDataException ex)
+                        {
+                            Log.RequestFormInvalidDataException(httpContext, ex);
+                            httpContext.Response.StatusCode = 400;
+                            return;
+                        }
```

Comment:
This is basically copy-pasta from the above block, with more specific log messages. Open to alternatives to make this more DRY and if it doesn't need the more-specific logging.

Reply:
This should be another lambda as you can't have both a JSON and form body.

Reply:
587d22f605194c763b80a81e06357fa46e6ed310